### PR TITLE
Added auto-optimization option for data loaders

### PR DESCRIPTION
**Pull request by avideci**
I enabled choosing the sweet spot for data loading processes count based the FPS that are loaded from disk, considering the current CPU utilization, trying different configurations.

The idea is to look at the hardware upon startup, or eventually on every epoch.
We should understand what's the best processing configuration and DDP for data loading, assuming other training procedures use other GPUs but they share CPUs with our process.

We try different constellations of data loading (instead of default data_loader_num_workers: int = 8).
You can try this feature by changing the default signature of the method (or pass True to the flag):
`
    def connect_dataset_interface(self, dataset_interface: DatasetInterface, data_loader_num_workers: int = 8,
                                  fine_tune=False):
`

## Current behavior
We use different kind of data_loader_num_workers all over the code, 8 being default in most places. some recipes change it to 4 / 32, without a good explanation (it is mostly dataset-dependent).

## The new behavior:
Speficying  fine_tune=True to sg_model's will ignore the data_loader_num_workers (even if it was specified in the recipes using Hydra yml configuration).
The algorithm will choose the number of worker that reached the highest utilization.
We can move it to a different function, I just wanted the code to be portable and it can be moved to any other class for test purposes.

We should try this on a busy machine, preferably a machine that runs other training procedures in parallel, to see the CPU and resource sharing.

## Some thoughts about the current configuration
We currently use data_loading_num_workers=8, which is not the sweet spot at most cases. I have seen 1 beating 8 many times - it can just hurt us, due to many OS paging operations and threading bottlenecks.
TLDR - For small datasets (CIFAR10 tested), num_workers=2 will be just as good as 8 for small datasets. Sometimes, 1 will be better than 2.

Outputs:
```text
Files already downloaded and verified
Files already downloaded and verified
Trying different data loader configuration (hardware-aware)...
Trying 1 data loaders...
1 data loaders yielded 128 samples in 10750.08ms...
FPS/Worker ratio: 83.98501032036165 Total FPS: 83.98501032036165
Trying 2 data loaders...
2 data loaders yielded 128 samples in 10799.11ms...
FPS/Worker ratio: 42.18400939453204 Total FPS: 84.36801878906408
Trying 3 data loaders...
3 data loaders yielded 128 samples in 10718.65ms...
FPS/Worker ratio: 27.91315065364112 Total FPS: 83.73945196092336
Trying 4 data loaders...
4 data loaders yielded 128 samples in 10906.30ms...
FPS/Worker ratio: 21.30136902343338 Total FPS: 85.20547609373352
Trying 5 data loaders...
5 data loaders yielded 128 samples in 10893.89ms...
FPS/Worker ratio: 17.021708420315917 Total FPS: 85.10854210157959
Trying 6 data loaders...
6 data loaders yielded 128 samples in 10847.65ms...
FPS/Worker ratio: 14.124549915360754 Total FPS: 84.74729949216453
Trying 7 data loaders...
7 data loaders yielded 128 samples in 10924.26ms...
FPS/Worker ratio: 12.19225172768087 Total FPS: 85.3457620937661
Trying 8 data loaders...
8 data loaders yielded 128 samples in 10906.46ms...
FPS/Worker ratio: 10.65083735839778 Total FPS: 85.20669886718224
Trying 9 data loaders...
9 data loaders yielded 128 samples in 11055.28ms...
FPS/Worker ratio: 9.59659831076149 Total FPS: 86.36938479685341
Trying 10 data loaders...
10 data loaders yielded 128 samples in 10784.65ms...
FPS/Worker ratio: 8.425506231253621 Total FPS: 84.25506231253621
Trying 11 data loaders...
11 data loaders yielded 128 samples in 10846.65ms...
FPS/Worker ratio: 7.70358631676903 Total FPS: 84.73944948445933
Trying 12 data loaders...
12 data loaders yielded 128 samples in 10923.51ms...
FPS/Worker ratio: 7.111658591156811 Total FPS: 85.33990309388173
Trying 13 data loaders...
13 data loaders yielded 128 samples in 10980.10ms...
FPS/Worker ratio: 6.598616255412428 Total FPS: 85.78201132036156
Trying 14 data loaders...
14 data loaders yielded 128 samples in 12460.97ms...
FPS/Worker ratio: 6.95366869587945 Total FPS: 97.35136174231229
Trying 15 data loaders...
15 data loaders yielded 128 samples in 12525.61ms...
FPS/Worker ratio: 6.523753192186632 Total FPS: 97.85629788279948
Trying 16 data loaders...
16 data loaders yielded 128 samples in 13186.65ms...
FPS/Worker ratio: 6.4387941850583275 Total FPS: 103.02070696093324
Data loading matrix: {1: 83.98501032036165, 2: 84.36801878906408, 3: 83.73945196092336, 4: 85.20547609373352, 5: 85.10854210157959, 6: 84.74729949216453, 7: 85.3457620937661, 8: 85.20669886718224, 9: 86.36938479685341, 10: 84.25506231253621, 11: 84.73944948445933, 12: 85.33990309388173, 13: 85.78201132036156, 14: 97.35136174231229, 15: 97.85629788279948, 16: 103.02070696093324}
Choosing num_processes=16 for data loading because it yields the best results at the moment.
```

```python
import pandas as pd
import matplotlib.pyplot as plt

d = {1: 83.98501032036165, 2: 84.36801878906408, 3: 83.73945196092336, 4: 85.20547609373352, 5: 85.10854210157959, 6: 84.74729949216453, 7: 85.3457620937661, 8: 85.20669886718224, 9: 86.36938479685341, 10: 84.25506231253621, 11: 84.73944948445933, 12: 85.33990309388173, 13: 85.78201132036156, 14: 97.35136174231229, 15: 97.85629788279948, 16: 103.02070696093324}
pd.DataFrame([d], index=None).plot(kind='bar')
plt.show()
```
image

As you can see, num_workers=16 was chosen but it is not necessarily the best. It was specific for that moment, considering other running tasks.

### CODE REVIEW

File: src/super_gradients/training/sg_model/sg_model.py [hunk 1]
```diff
@@ -200,13 +202,67 @@ def __init__(self, experiment_name: str, device: str = None, multi_gpu: Union[Mu
...
         :param data_loader_num_workers: The number of threads to initialize the Data Loaders with
             The dataset to be connected
+
+        :param fine_tune: This flag will ignore data_loader_num_workers.
+            When set to True, the program will search for a 'sweet spot' in terms of number of data loading processes.
+            The constellation that yields the highest data loading speed will be chosen.
+            This flag is useful when training multiple models over the same CPUs (same machine).
         """
         self.dataset_interface = dataset_interface
+        if fine_tune:
```
Comment (shaydeci) [inline src/super_gradients/training/sg_model/sg_model.py:218]: Can we move extract this code block to a method in utils ? Does not fell like this logic should be inside connect_dataset_interface.

**Review by shaydeci — CHANGES_REQUESTED** [review]
inline

File: src/super_gradients/training/sg_model/sg_model.py (hunk 1 above)
Comment (avideci) [reply src/super_gradients/training/sg_model/sg_model.py:218]: Definitely, it was just here for easy context so you can understand it inline.
Will move ti another file after we agree on the logic.

File: src/super_gradients/training/sg_model/sg_model.py [hunk 2]
```diff
@@ -200,13 +202,67 @@ def __init__(self, experiment_name: str, device: str = None, multi_gpu: Union[Mu
...
+                _train_loader, _valid_loader, _test_loader, _classes = \
+                    self.dataset_interface.get_data_loaders(batch_size_factor=self.num_devices,
+                                                            num_workers=potential_num_workers,
+                                                            distributed_sampler=self.multi_gpu == MultiGPUMode.DISTRIBUTED_DATA_PARALLEL)
+
+                # Warmup
+                [iter(_train_loader) for _ in range(WARMUP_CALLS)]
+
+                samples_loaded = 0
+                start_load_sample = time.perf_counter()
```
[the remark this answers is not in the document]
Comment (avideci) [reply src/super_gradients/training/sg_model/sg_model.py:241]: They all finish after the list comprehension, don't they?
This step will always be sync, and will run on the main thread, correct me if I am worong

File: src/super_gradients/training/sg_model/sg_model.py [hunk 3]
```diff
@@ -200,13 +202,67 @@ def __init__(self, experiment_name: str, device: str = None, multi_gpu: Union[Mu
         self.train_metrics, self.valid_metrics = default_train_metrics, default_valid_metrics
         self.loss_logging_items_names = default_loss_logging_items_names
 
-    def connect_dataset_interface(self, dataset_interface: DatasetInterface, data_loader_num_workers: int = 8):
+    def connect_dataset_interface(self, dataset_interface: DatasetInterface, data_loader_num_workers: int = 8,
+                                  fine_tune=False):
```
[the remark this answers is not in the document]
Comment (avideci) [reply src/super_gradients/training/sg_model/sg_model.py:206]: I agree, find_optimal_data_loading_num_workers() would be better, 
Do you think an argument in this function is the right place?

**Comment by avideci** [comment]
Did you manage to run it on a machine with running workloads? @shaydeci 
I tested it on CIFAR10 but we need larger datasets as well (ImageNet, COCO...)
We must see if the number will change for a clean environment vs. busy environment with utilization.

**Comment by shaydeci** [comment]
Im closing this down for now, as it has been stale for a while with no significant advantage at the moment.
